File: sdks/node/src/path-normalizer.ts
/**
 * Normalize a URL path by replacing dynamic segments with parameter placeholders.
 *
 * If a framework route template is available (e.g., Express `req.route?.path`),
 * use that directly — it already has :param placeholders.
 *
 * Otherwise, apply heuristic normalization:
 *   /users/123         → /users/:id
 *   /orders/abc-def    → /orders/:id
 *   /v1/items/42/reviews/7 → /v1/items/:id/reviews/:id
 *
 * UUIDs, numeric IDs, and alphanumeric slugs are detected and replaced.
 */

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
const NUMERIC_RE = /^\d+$/;
const MONGO_ID_RE = /^[0-9a-f]{24}$/i;
// Version-like segments: v1, v2, v10, etc. — should NOT be normalized
const VERSION_RE = /^v\d+$/i;

export function normalizePath(
  actualPath: string,
  routeTemplate?: string,
): string {
  // Prefer the framework's route template if available
  if (routeTemplate) {
    return routeTemplate;
  }

  // Heuristic: replace dynamic-looking segments
  const segments = actualPath.split('/');
  const normalized = segments.map((seg) => {
    if (!seg) return seg;
    // Preserve version segments (v1, v2, v10)
    if (VERSION_RE.test(seg)) return seg;
    if (UUID_RE.test(seg)) return ':id';
    if (NUMERIC_RE.test(seg)) return ':id';
    if (MONGO_ID_RE.test(seg)) return ':id';
    // Alphanumeric tokens > 4 chars that mix letters and digits look like IDs
    if (seg.length > 4 && seg.length <= 12 && /^[a-zA-Z0-9_-]+$/.test(seg) && /\d/.test(seg) && /[a-zA-Z]/.test(seg)) {
      return ':id';
    }
    return seg;
  });

  return normalized.join('/');
}

File: sdks/python/aforo/path_normalizer.py
"""URL path normalization — replaces dynamic segments with :id placeholders."""

from __future__ import annotations

import re
from typing import Optional

_UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.I
)
_NUMERIC_RE = re.compile(r"^\d+$")
_MONGO_RE = re.compile(r"^[0-9a-f]{24}$", re.I)
_VERSION_RE = re.compile(r"^v\d+$", re.I)


def normalize_path(actual_path: str, route_template: Optional[str] = None) -> str:
    """Normalize a URL path by replacing dynamic segments with ``:id``.

    If *route_template* is provided (e.g. from the framework router),
    it is returned directly.
    """
    if route_template:
        return route_template

    segments = actual_path.split("/")
    normalized = []
    for seg in segments:
        if not seg:
            normalized.append(seg)
            continue
        if _VERSION_RE.match(seg):
            normalized.append(seg)
        elif _UUID_RE.match(seg) or _NUMERIC_RE.match(seg) or _MONGO_RE.match(seg):
            normalized.append(":id")
        elif (
            len(seg) > 4
            and len(seg) <= 12
            and re.match(r"^[a-zA-Z0-9_-]+$", seg)
            and re.search(r"\d", seg)
            and re.search(r"[a-zA-Z]", seg)
        ):
            normalized.append(":id")
        else:
            normalized.append(seg)

    return "/".join(normalized)
